src/olmo_core/train/trainer.py

```diff
@@ -529,15 +529,21 @@ def training_progress(self) -> TrainingProgress:
             time_remaining=time_remaining,
         )
 
-    def cancel_run(self, reason: str):
+    def cancel_run(self, reason: str, no_sync: bool = False):
         """
         Mark the run canceled.
 
         :param reason: The reason for canceling.
+        :param no_sync: Set this to ``True`` only if you're calling this from all ranks at the same
+            time, otherwise you'll get a distributed deadlock.
         """
-        #  self._canceled = True  # NOTE: important not to set this!! Leads to distributed hang.
         self._canceling_rank = get_rank()
         self._cancel_reason = reason
+        if no_sync:
+            self._canceled = (
+                True  # NOTE: important not to set this otherwise!! Leads to distributed hang.
+            )
```

Comment:
This is confusing. The comment says to not set this, but it is being set?

Reply:
Done